File: src/lib/sources/hackernews.ts
import { RawSearchResult } from '@/types';

// HackerNews Algolia API - 免费，无需 key，实时性好
const HN_API = 'https://hn.algolia.com/api/v1';

export async function searchHackerNews(keyword: string): Promise<RawSearchResult[]> {
  try {
    const query = encodeURIComponent(keyword);
    // 搜索最近 7 天内的帖子，按时间排序
    const url = `${HN_API}/search_by_date?query=${query}&tags=(story,show_hn)&numericFilters=created_at_i>${Math.floor(Date.now() / 1000) - 7 * 86400}&hitsPerPage=10`;

    const res = await fetch(url, {
      signal: AbortSignal.timeout(8000),
    });

    if (!res.ok) return [];

    const data = await res.json();
    const hits = data.hits || [];

    return hits
      .filter((h: HNHit) => h.title && h.points >= 5)
      .map((h: HNHit) => ({
        title: h.title,
        url: h.url || `https://news.ycombinator.com/item?id=${h.objectID}`,
        snippet: `HN ${h.points} points · ${h.num_comments} comments · by ${h.author}`,
        source_type: 'hackernews' as const,
        author: h.author,
        published_at: h.created_at,
      }));
  } catch (e) {
    console.error('[hackernews] Search failed:', e);
    return [];
  }
}

// 获取 HN 热门故事（不依赖关键词，直接拿热榜）
export async function getHackerNewsTop(keyword: string): Promise<RawSearchResult[]> {
  try {
    // 先搜索相关内容
    const searchResults = await searchHackerNews(keyword);

    // 再从热门故事里过滤相关的
    const topRes = await fetch(`${HN_API}/search?query=${encodeURIComponent(keyword)}&tags=front_page&hitsPerPage=5`, {
      signal: AbortSignal.timeout(8000),
    });

    if (!topRes.ok) return searchResults;

    const topData = await topRes.json();
    const keywordLower = keyword.toLowerCase();
    const topHits = (topData.hits || [])
      .filter((h: HNHit) => h.title?.toLowerCase().includes(keywordLower))
      .map((h: HNHit) => ({
        title: h.title,
        url: h.url || `https://news.ycombinator.com/item?id=${h.objectID}`,
        snippet: `🔥 HN Front Page · ${h.points} points · ${h.num_comments} comments`,
        source_type: 'hackernews' as const,
        author: h.author,
        published_at: h.created_at,
      }));

    return [...topHits, ...searchResults].slice(0, 10);
  } catch (e) {
    console.error('[hackernews] Top fetch failed:', e);
    return [];
  }
}

interface HNHit {
  objectID: string;
  title: string;
  url?: string;
  points: number;
  num_comments: number;
  author: string;
  created_at: string;
}

File: agent-skills/hot-monitor/scripts/sources/hackernews.py
"""Ported from src/lib/sources/hackernews.ts"""

from __future__ import annotations

import time
from urllib.parse import quote

import httpx

from models import RawSearchResult

HN_API = "https://hn.algolia.com/api/v1"


async def search_hackernews(keyword: str) -> list[RawSearchResult]:
    try:
        since = int(time.time()) - 7 * 86400
        url = (
            f"{HN_API}/search_by_date?query={quote(keyword)}"
            f"&tags=(story,show_hn)&numericFilters=created_at_i>{since}&hitsPerPage=10"
        )
        async with httpx.AsyncClient(timeout=8.0) as client:
            res = await client.get(url)
        if res.status_code != 200:
            return []

        data = res.json()
        hits = data.get("hits") or []
        results: list[RawSearchResult] = []

        for h in hits:
            if not h.get("title") or (h.get("points") or 0) < 5:
                continue
            results.append(
                {
                    "title": h["title"],
                    "url": h.get("url") or f"https://news.ycombinator.com/item?id={h['objectID']}",
                    "snippet": (
                        f"HN {h.get('points', 0)} points · "
                        f"{h.get('num_comments', 0)} comments · by {h.get('author', '')}"
                    ),
                    "source_type": "hackernews",
                    "author": h.get("author"),
                    "published_at": h.get("created_at"),
                }
            )
        return results
    except Exception as e:
        print(f"[hackernews] Search failed: {e}", flush=True)
        return []


async def get_hackernews_top(keyword: str) -> list[RawSearchResult]:
    try:
        search_results = await search_hackernews(keyword)

        url = f"{HN_API}/search?query={quote(keyword)}&tags=front_page&hitsPerPage=5"
        async with httpx.AsyncClient(timeout=8.0) as client:
            top_res = await client.get(url)
        if top_res.status_code != 200:
            return search_results

        top_data = top_res.json()
        keyword_lower = keyword.lower()
        top_hits: list[RawSearchResult] = []

        for h in top_data.get("hits") or []:
            title = h.get("title") or ""
            if keyword_lower not in title.lower():
                continue
            top_hits.append(
                {
                    "title": title,
                    "url": h.get("url") or f"https://news.ycombinator.com/item?id={h['objectID']}",
                    "snippet": (
                        f"🔥 HN Front Page · {h.get('points', 0)} points · "
                        f"{h.get('num_comments', 0)} comments"
                    ),
                    "source_type": "hackernews",
                    "author": h.get("author"),
                    "published_at": h.get("created_at"),
                }
            )

        return (top_hits + search_results)[:10]
    except Exception as e:
        print(f"[hackernews] Top fetch failed: {e}", flush=True)
        return []
